지윤/chapter2/item16.md

```diff
@@ -0,0 +1,47 @@
+#### 아이템16. number 인덱스 시그니처보다는 Array, 튜플, ArrayLike를 사용하자.
+
+자바스크립트에서의 객체란, 키/값 쌍의 모음이다. 키는 보통 문자열이고, 값는 무엇이든지 될 수 있다. 복잡한 객체나, 숫자를 키로 사용하려하면, 자바스크립트는 해당 객체나 숫자를 문자열로 변환해 키로 사용한다. 배열 또한 객체이기에, 배열의 인덱스 또한 키라, 문자열로 취급된다. 
+```javascript
+const numbers = [1,2,3];
+console.log(numbers['2']) // 3
+console.log(Object.keys(numbers)) // ['0', '1', '2']
+```
+
+배열의 인덱스에 문자열이 들어가는타입스크립트는 숫자키를 허용한다. 런타임 시에 타입 정보가 제거되기에, 런타임에는 문자열로 변환되기는 하지만, 타입 체크 시점에 오류를 잡을 수 있어 유용하다.
+```typescript
+// ??? 하지만 에러가 안난다..? 왜지?
+const numbers = [1,2,3]
+const a = numbers['1']
+console.log(numbers['1'])
+```
```

Comment:
https://stackoverflow.com/questions/60912874/why-dont-arrays-have-string-index-signatures

찾아봤는데, IndexType<number>가 IndexType<string>의 서브 타입이라고는 합니다. 정확한건 좀 더 찾아봐야할 듯해요~

Reply:
요 부분 제 쪽에 리뷰 달아주셔서 거기에 추가 코멘트 달았는데! 버전 차이인 듯 합니다! 기존에는 좁혔던 타입을 넓히면서 그 차이에서 오는 상황 같습니다!
그리고 해당 코드 예제도 업뎃이 19년도 말에서 안되어있기에 그런 듯 합니다!

Reply:
오 미쳤당 제가 chatGPT한테 물어봤씁니다 명확하게 답변을 받았씁니다.

In JavaScript, arrays are a type of object, and the keys of an object can be either strings or numbers. When you access an array with a string index, JavaScript will try to convert the string to a number and use that as the index.

In TypeScript, when you define an array like let x: number[] = [1, 2, 3], it infers the index as number type. But when you use a string index to access the array, `TypeScript does not throw an error because it allows to use string index with the array if the array is defined without explicit index type.`

However, `you can use the keyof operator to restrict the index to a specific type.` For example, you can use let x: { [index: number]: number } = [1, 2, 3] to restrict the index to a number type.